slack-plugin/src/main/kotlin/slack/gradle/settings/SgpSettingsPlugin.kt

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2023 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package slack.gradle.settings
+
+import javax.inject.Inject
+import org.gradle.api.Action
+import org.gradle.api.Plugin
+import org.gradle.api.artifacts.VersionCatalog
+import org.gradle.api.initialization.Settings
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.plugins.ExtensionAware
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
+import org.gradle.api.provider.ProviderFactory
+import org.gradle.kotlin.dsl.apply
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.newInstance
+import org.gradle.kotlin.dsl.property
+import slack.gradle.SlackExtensionMarker
+import slack.gradle.SlackVersions
+import slack.gradle.agp.AgpHandler
+import slack.gradle.agp.AgpHandlerFactory
+import slack.gradle.agp.getVersionsCatalog
+import slack.gradle.isRootProject
+import slack.gradle.settings.SgpSettingsPlugin.Companion.AGP_HANDLER_EXTENSION_KEY
+import slack.gradle.settings.SgpSettingsPlugin.Companion.SGP_VERSIONS_EXTENSION_KEY
+import slack.gradle.tomlKey
+
+public class SgpSettingsPlugin : Plugin<Settings> {
+  override fun apply(settings: Settings) {
+    val catalog = settings.getVersionsCatalog()
+    val extension = settings.extensions.create<SgpSettingsExtension>("slack", catalog)
+
+    val agpHandlerFactory = AgpHandlerFactory.load()
+    agpHandlerFactory.createSettingsHandler().apply(settings)
+    val handler = agpHandlerFactory.createHandler()
+
+    lateinit var versions: SlackVersions
+    settings.gradle.settingsEvaluated { versions = extension.versions.finalize() }
+
+    // as Project objects cannot query for the Settings object (and its extensions), we
+    // deposit the handler instance into each project using the extra Properties.
+    settings.gradle.beforeProject {
+      // Gradle does this immensely weird thing where it considers all intermediary folders in
+      // between projects to _also_ be projects. We skip those in configuration.
+      if (!buildFile.exists()) return@beforeProject
+
+      extensions.extraProperties[SGP_VERSIONS_EXTENSION_KEY] = versions
+      extensions.extraProperties[AGP_HANDLER_EXTENSION_KEY] = handler
+
+      // Apply base plugin to all subprojects
```

Comment:
I'm curious about how this works in practice for you. In our set of convention plugins, we also have a base plugin (because of course), but we also have baked in the assumption that the base plugin is applied first, and then subsequent plugins can then count on that being true. I assume that `beforeProject {}` ensures the base plugin is applied first, right? And just thinking out loud... I guess the fact that these plugins are applied via the settings script means they're loaded in the settings classloader, whereas other plugins might be loaded in different (probably child) classloaders. I guess everything "just works"? But I always wonder a bit with gradle's crazy classloader hierarchy.

Reply:
SGP is more or less order-agnostic when it comes to other plugins, so this wouldn't change anything for us at least!